File: 序号/19.删除链表的倒数第N个节点.cpp
class Solution {
 public:
  /** 删除倒数第n个节点 */
  ListNode* removeNthFromEnd(ListNode* head, int n) {
    // 滑动窗口
    ListNode *fakeHead = new ListNode(0,head);
    ListNode *pop=fakeHead;
    ListNode *p=fakeHead->next;
    deque<ListNode *> q;
    while(p){
      if(q.size()<n) q.emplace_back(p);
      else {
        pop=q.front();
        q.pop_front();
        q.emplace_back(p);
      }
      p=p->next;
    }
    if (n > 1) pop->next = q[1];
    else pop->next = nullptr;
    ListNode * ans = fakeHead->next;
    delete fakeHead;
    return ans;
  }
};

File: 序号/19.删除链表的倒数第N个节点.py
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(next=head)
        pre = dummy
        p = head
        q = []
        while p:
            if (len(q) < n):
                q.append(p)
            else:
                pre = q.pop(0)
                q.append(p)
            p = p.next
        if (n == 1):
            pre.next = None
        else:
            pre.next = q[1]
        return dummy.next